src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeQueries.java

```diff
@@ -0,0 +1,181 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.disk.v1.bbtree;
+
+import java.nio.ByteBuffer;
+
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.index.sai.utils.TypeUtil;
+import org.apache.lucene.index.PointValues.Relation;
+import org.apache.lucene.util.FutureArrays;
+
+public class BlockBalancedTreeQueries
+{
+    private static final BlockBalancedTreeReader.IntersectVisitor MATCH_ALL = new BlockBalancedTreeReader.IntersectVisitor()
+    {
+        @Override
+        public boolean visit(byte[] packedValue)
+        {
+            return true;
+        }
+
+        @Override
+        public Relation compare(byte[] minPackedValue, byte[] maxPackedValue)
+        {
+            return Relation.CELL_INSIDE_QUERY;
+        }
+    };
+
+    public static BlockBalancedTreeReader.IntersectVisitor balancedTreeQueryFrom(Expression expression, int bytesPerValue)
+    {
+        if (expression.lower == null && expression.upper == null)
+        {
+            return MATCH_ALL;
+        }
+
+        Bound lower = null ;
+        if (expression.lower != null)
+        {
+            final byte[] lowerBound = toComparableBytes(bytesPerValue, expression.lower.value.encoded, expression.validator);
+            lower = new Bound(lowerBound, !expression.lower.inclusive);
+        }
+
+        Bound upper = null;
+        if (expression.upper != null)
+        {
+            final byte[] upperBound = toComparableBytes(bytesPerValue, expression.upper.value.encoded, expression.validator);
+            upper = new Bound(upperBound, !expression.upper.inclusive);
+        }
+
+        return new RangeQueryVisitor(bytesPerValue, lower, upper);
+    }
+
+    private static byte[] toComparableBytes(int bytesPerDim, ByteBuffer value, AbstractType<?> type)
+    {
+        byte[] buffer = new byte[TypeUtil.fixedSizeOf(type)];
+        assert buffer.length == bytesPerDim;
+        TypeUtil.toComparableBytes(value, type, buffer);
+        return buffer;
+    }
+
+    private static abstract class RangeQuery implements BlockBalancedTreeReader.IntersectVisitor
+    {
+        final int bytesPerValue;
+
+        RangeQuery(int bytesPerValue)
+        {
+            this.bytesPerValue = bytesPerValue;
+        }
+
+        int compareUnsigned(byte[] packedValue, Bound bound)
```

Comment:
It feels like the code that uses this would be clearer if we...

1.) Removed `compareUnsigned()` here.

2.) Changed `Bound` to look like...

```
boolean smallerThan(byte[] packedValue)
        {
            int cmp = compareTo(packedValue, bound.length);
            return cmp < 0 || (cmp == 0 && exclusive);
        }

        boolean greaterThan(byte[] packedValue)
        {
            int cmp = compareTo(packedValue, bound.length);
            return cmp > 0 || (cmp == 0 && exclusive);
        }

        private int compareTo(byte[] packedValue, int length)
        {
            assert packedValue.length == length : "WTF";
            return FutureArrays.compareUnsigned(bound, 0, length, packedValue, 0, length);
        }
```

3.) Changed usages to look like `if (lower.greaterThan(maxPackedValue))`